File: core/src/main/java/com/gumse/gui/Basics/Tabs.cpp
#include "Tabs.h"
#include "../Font/FontManager.h"
#include "TextBox.h"

#include <System/MemoryManagement.h>
#include <Essentials/Tools.h>
#include <OpenGL/WrapperFunctions.h>

Tabs::Tabs(vec2 pos, vec2 size, ivec2 tabsize)
{
    this->sType = "Tabs";
    this->vPos = pos;
    this->vSize = size;
    this->vTabSize = tabsize;
    this->v4ActiveColor = vec4(0.5,0.5,0.5, 1.0);
    this->v4InactiveColor = vec4(0.4,0.4,0.4, 1.0);
    this->pActiveTab = nullptr;

    pBackground = new Box(ivec2(0,0), ivec2(100, tabsize.y));
    pBackground->setColor(vec4(0.25,0.25,0.25,1));
    pBackground->setSizeInPercent(true, false);
    pBackground->setOrigin(ivec2(0, tabsize.y));
    addElement(pBackground);

    vMargin.y = -tabsize.y;
    vOrigin.y = -tabsize.y;

    sfOffset.setMin(0.0f);


    resize();
    reposition();
}

Tabs::~Tabs()
{
    pBackground->destroyChildren();
    Gum::_delete(pBackground);
}

void Tabs::update()
{
    for(size_t i = 0; i < vTabs.size(); i++)
    {
        if(vTabs[i]->isClicked())
        {
            if(pActiveTab != nullptr)
            {
                pActiveTab->setColor(v4InactiveColor);
                pActiveTab->hideChildren(true);
            }
            vTabs[i]->setColor(v4ActiveColor);
            vTabs[i]->hideChildren(false);
            pActiveTab = vTabs[i];
        }
    }

    TextBox* lastTab = vTabs[vTabs.size() - 1];
    if(lastTab->getPosition().x + lastTab->getSize().x > vActualSize.x &&
        Tools::checkPointInBox(Gum::Window::CurrentlyBoundWindow->getMouse()->getPosition(), bbox2i(vActualPos - ivec2(0, vTabSize.y), ivec2(vActualSize.x, vTabSize.y))))
    {
        sfOffset.setMax(lastTab->getRelativePosition().x + lastTab->getSize().x - vActualSize.x + lastTab->getOrigin().x);
        sfOffset.increaseTarget(-50 * Gum::Window::CurrentlyBoundWindow->getMouse()->getMouseWheelState());
    }

    if(sfOffset.update())
    {
        for(unsigned int i = 0; i < pBackground->numChildren(); i++)
        {
            RenderGUI* elem = pBackground->getChild(i);
            elem->setOrigin(ivec2(sfOffset.get(), 0));
        }
    }
    updatechildren();
}

bool Tabs::updateToolTip()
{
    if(isMouseInside())
    {
        for(unsigned int i = 0; i < pActiveTab->numChildren(); i++) 
        {
            bool showToolTip = true;
            if(pActiveTab->getChild(i)->updateToolTip())
            {
                showToolTip = false;
            }

            //if(showToolTip)
            //    GumEngine::GUIS->showToolTip(sToolTip);
        }
        return true;
    }
    return false;
}


void Tabs::render()
{
    glEnable(GL_SCISSOR_TEST);
    gumScissor(bbox2i(ivec2(vActualPos.x, Gum::Window::CurrentlyBoundWindow->getSize().y - vActualPos.y - vActualSize.y + vTabSize.y), vActualSize));
    pBackground->render();
    glDisable(GL_SCISSOR_TEST);
}

void Tabs::addGUIToTab(RenderGUI *gui, std::string tabName)
{
    TextBox* tab = getTab(tabName);
    if(tab != nullptr)
    {
        tab->addGUI(gui);
        gui->setParent(this); //Overwrite parent
        gui->reposition();
        gui->resize();
    }
}

void Tabs::addTab(std::string name, bool active)
{
    float width = 0.0f;
    for(TextBox* tab : vTabs)
        width += tab->getSize().x;

    TextBox* tab = new TextBox(name, Gum::GUI::Fonts->getDefaultFont(), ivec2(width, 0), vTabSize);
    tab->setTextSize(vTabSize.y - 10);
    if(active)
    {
        tab->setColor(v4ActiveColor);
        tab->hideChildren(false);
        pActiveTab = tab;
    }
    else
    {
        tab->setColor(v4InactiveColor);
        tab->hideChildren(true);
    }
    tab->setSize(ivec2(tab->getTextSize().x + 30, vTabSize.y));
    tab->setTextOffset(ivec2(5, 3));
    pBackground->addGUI(tab);
    vTabs.push_back(tab);
}

void Tabs::setActiveTab(std::string tabname)
{
    TextBox* pFoundTab = getTab(tabname);
    if(pFoundTab != nullptr)
        this->pActiveTab = pFoundTab;
}

bool Tabs::isActiveTab(std::string tabname)
{
    TextBox* pFoundTab = getTab(tabname);
    return pFoundTab != nullptr && pActiveTab == pFoundTab;
}

TextBox* Tabs::getTab(std::string name)
{
    for(size_t i = 0; i < vTabs.size(); i++)
        if(vTabs[i]->getTitle().toString() == name)
            return vTabs[i];
    return nullptr;
}

TextBox* Tabs::getActiveTab()            { return pActiveTab; }
unsigned int Tabs::numTabs()             { return vTabs.size(); }

Tabs* Tabs::createFromXMLNode(XMLNode* node)
{
    ivec2 tabSize = node->mAttributes["tabsize"] != "" ? (ivec2)Tools::StringToVec2(node->mAttributes["tabsize"]) : ivec2(100, 30);
    return new Tabs(ivec2(0,0), ivec2(1,1), tabSize);
}

File: core/src/main/java/com/gumse/gui/Basics/Tabs.java
package com.gumse.gui.Basics;

import org.lwjgl.opengl.GL11;

import com.gumse.PostProcessing.Framebuffer;
import com.gumse.basics.SmoothFloat;
import com.gumse.gui.GUI;
import com.gumse.gui.Font.FontManager;
import com.gumse.gui.Primitives.Box;
import com.gumse.gui.Primitives.RenderGUI;
import com.gumse.gui.XML.XMLGUI.XMLGUICreator;
import com.gumse.maths.*;
import com.gumse.system.Window;
import com.gumse.system.filesystem.XML.XMLNode;
import com.gumse.system.io.Mouse;
import com.gumse.tools.Output;

public class Tabs extends RenderGUI
{
    private RenderGUI pActiveTab;
    private Box pBackground;
    private ivec2 vTabSize;
    private SmoothFloat sfOffset;


    public Tabs(ivec2 pos, ivec2 size, ivec2 tabsize)
    {
        this.sType = "Tabs";
        this.vPos.set(pos);
        this.vSize.set(size);
        this.vTabSize = tabsize;
        this.pActiveTab = null;
        this.sfOffset = new SmoothFloat();

        pBackground = new Box(new ivec2(0,0), new ivec2(100, tabsize.y));
        pBackground.setColor(GUI.getTheme().secondaryColor);
        pBackground.setSizeInPercent(true, false);
        pBackground.setOrigin(new ivec2(0, tabsize.y));
        addElement(pBackground);

        vMargin.y = -tabsize.y;
        vOrigin.y = -tabsize.y;

        sfOffset.setMin(0.0f);


        resize();
        reposition();
    }

    public void cleanup()
    {
        pBackground.destroyChildren();
    }

    public void updateextra()
    {
        if(pBackground.numChildren() > 0)
        {
            TextBox lastTab = (TextBox)pBackground.getChild(pBackground.numChildren() - 1);
            if(pBackground.isMouseInsideSkipChildren())
            {
                sfOffset.setMax(lastTab.getRelativePosition().x + lastTab.getSize().x - vActualSize.x + lastTab.getOrigin().x);
                sfOffset.increaseTarget(-50 * Window.CurrentlyBoundWindow.getMouse().getMouseWheelState());
            }

            if(sfOffset.update())
            {
                for(int i = 0; i < pBackground.numChildren(); i++)
                {
                    RenderGUI elem = pBackground.getChild(i);
                    elem.setOrigin(new ivec2((int)sfOffset.get(), 0));
                }
            }   
        }
    }

    @Override
    protected void updateOnThemeChange() 
    {
        pBackground.setColor(GUI.getTheme().secondaryColor);
        for(int i = 0; i < vElements.size(); i++)
            vElements.get(i).setColor(GUI.getTheme().primaryColorShade);

        if(getTab(pActiveTab.getTitle()) != null)
            getTab(pActiveTab.getTitle()).setColor(GUI.getTheme().primaryColorShade);
    }

    public void renderextra()
    {
        GL11.glEnable(GL11.GL_SCISSOR_TEST);
        GL11.glScissor(vActualPos.x, Framebuffer.CurrentlyBoundFramebuffer.getSize().y - vActualPos.y - vActualSize.y + vTabSize.y, vActualSize.x, vActualSize.y);
        renderchildren();
        GL11.glDisable(GL11.GL_SCISSOR_TEST);

    }

    public void addGUIToTab(RenderGUI gui, String tabName)
    {
        RenderGUI tab = getTabContent(tabName);
        if(tab != null)
            tab.addGUI(gui);
    }

    public void addTab(final String name, final boolean active)
    {
        RenderGUI tabContent = new RenderGUI();
        tabContent.setTitle(name);
        tabContent.setSize(new ivec2(100, 100));
        tabContent.setSizeInPercent(true, true);
        addElement(tabContent);

        int xoffset = 0;
        for(RenderGUI tab : pBackground.getChildren())
            xoffset += tab.getSize().x;

        TextBox tabSelector = new TextBox(name, FontManager.getInstance().getDefaultFont(), new ivec2(xoffset, 0), vTabSize);
        tabSelector.setSize(new ivec2(tabSelector.getTextSize().x + 30, vTabSize.y));
        tabSelector.setCornerRadius(new vec4(0,0,0,0));
        tabSelector.onClick((RenderGUI thisselector) -> {
            setActiveTab(name);
        });
        tabSelector.onHover(null, Mouse.GUM_CURSOR_HAND);
        pBackground.addGUI(tabSelector);

        if(active)
            setActiveTab(name);
    }

    public void setActiveTab(String tabname)
    {
        TextBox tab = getTab(tabname);
        RenderGUI content = getTabContent(tabname);
        if(tab != null)
        {
            //Disable all other tabs
            for(RenderGUI othertab : pBackground.getChildren())
                othertab.setColor(GUI.getTheme().primaryColor);

            for(RenderGUI othercontent : vElements)
            {
                if(!othercontent.getType().equals("box"))
                    othercontent.hide(true);
            }

            tab.setColor(GUI.getTheme().primaryColorShade);
            this.pActiveTab = content;
            content.hide(false);
        }
    }

    public boolean isActiveTab(String tabname)
    {
        RenderGUI pFoundTab = getTabContent(tabname);
        return pFoundTab != null && pActiveTab == pFoundTab;
    }

    private TextBox getTab(String name)
    {
        for(RenderGUI child : pBackground.getChildren())
        {
            if(child.getTitle().equals(name))
                return (TextBox)child;
        }
        return null;
    }

    public RenderGUI getTabContent(String name)
    {
        for(RenderGUI child : vElements)
        {
            if(child.getTitle().equals(name))
                return child;
        }
        return null;
    }

    public RenderGUI getActiveTabContent() { return pActiveTab; }
    public int numTabs()                   { return vElements.size(); }


    public static XMLGUICreator createFromXMLNode() 
    {
        return (XMLNode node) -> { 
            ivec2 tabSize = node.getIvec2Attribute("tabsize", new ivec2(100, 30));
            return new Tabs(new ivec2(0,0), new ivec2(1,1), tabSize);
        };
    };
};